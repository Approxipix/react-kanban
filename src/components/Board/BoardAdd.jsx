import React, { Component } from 'react';
import { connect } from 'react-redux';
import { history } from "../../redux/store";
import { bindActionCreators } from 'redux';
import { addBoard } from '../../redux/boardReducer/actions/actions';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Actions, Input, SubmitButton, CancelButton } from '../BaseComponent';
import ClickOutside from "../ClickOutside/ClickOutside";
import styled from 'styled-components';
import PropTypes from "prop-types";

const AddButton = styled.button`
  font-size: 1rem; 
  color: #172b4d;
  font-weight: 600;
  background-color: #dadde3;
  border-radius: .4rem
  transition: background-color .2s ease-in;
  cursor: pointer;
  &:hover {
    background-color: #c8cace;
  }
`;
AddButton.displayName = 'AddButton';

const Container = styled.div`
  height: 100%;
  padding: 1rem .6rem;
  background-color: #dadde3;
  border-radius: .4rem
`;

const ColorPicker = styled.div`
  display: flex;
  padding: .25rem 0;
  margin-bottom: .5rem;
`;
ColorPicker.displayName = 'ColorPicker';

const ColorPickerItem = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  font-size: .7rem;
  color: #fff;
  background-color: ${props => props.color};
  border-radius: .4rem;
  cursor: pointer;
  &:not(:last-child) {
    margin-right: .5rem;
  }
  &:hover {
    box-shadow: inset 0 0 0 10rem rgba(0, 0, 0, .1)
  }
`;

class BoardAdd extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      boardTitle: '',
      boardColor: props.colors[0] || '',
    };
  }

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.toggleOpened();
    }
  };

  handleChange = (key, value) => {
    this.setState({
      ...this.state,
      [key]: value,
    });
  };

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      boardTitle: '',
    })
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { boardTitle } = this.state;
    if (!boardTitle) return;
    const boardId = Date.now().toString();
    this.props.actions.addBoard({
      boardTitle: boardTitle,
      boardColor: this.state.boardColor,
      boardId: boardId,
    });
    history.push(`/b/${boardId}`)
  };

  render = () => {
    const { colors } = this.props;
    const { isOpened, boardTitle, boardColor } = this.state;
    return isOpened ? (
      <ClickOutside toggleOpened={this.toggleOpened}>
        <Container color={boardColor}>
          <form onSubmit={this.handleSubmit}>
            <Input
              type="text"
              placeholder="Add board title"
              value={boardTitle}
              onKeyDown={this.handleKeyDown}
              onChange={(e) => this.handleChange('boardTitle', e.target.value)}
              spellCheck={false}
              autoFocus
            />
            <ColorPicker>
              {colors.map((color, index) => (
                <ColorPickerItem
                  key={index}
                  color={color}
                  onClick={() => this.handleChange('boardColor', color)}
                >
                  {boardColor === color && <FontAwesomeIcon icon="check"/>}
                </ColorPickerItem>
              ))}
            </ColorPicker>
            <Actions>
              <SubmitButton type='submit' disabled={!boardTitle}>
                Create board
              </SubmitButton>
              <CancelButton onClick={() => this.toggleOpened()}>
                <FontAwesomeIcon icon="times"/>
              </CancelButton>
            </Actions>
          </form>
        </Container>
      </ClickOutside>
    ) : (
      <AddButton onClick={() => this.toggleOpened()}>
        Create new board...
      </AddButton>
    );
  };
}

BoardAdd.defaultProps = {
  colors: [],
};

BoardAdd.propTypes = {
  colors: PropTypes.arrayOf(PropTypes.string).isRequired,
};

function mapStateToProps(state) {
  return {
    colors: state.rootReducer.colors,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addBoard: addBoard,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(BoardAdd);
