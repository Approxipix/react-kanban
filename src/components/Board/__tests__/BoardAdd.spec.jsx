import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import BoardAdd from '../BoardAdd';

const mockStore = configureStore([]);

describe('<BoardAdd>', () => {
  let store;
  let props;
  let state;
  let component;
  let mockAddBoard = jest.fn();

  beforeEach(() => {
    store = mockStore({
      rootReducer: {
        colors: ['#2E7EAF', '#00603d', '#D29034'],
      }
    });
    store.dispatch = mockAddBoard;
    state = {
      isOpened: false,
      boardTitle: '',
      boardColor: '',
    };

    const wrapper = shallow(<BoardAdd store={store} {...props} />);
    component = wrapper.find('BoardAdd').dive();
    component.setState({ ...state })
  });

  it('should render AddButton if isOpened is false', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render form if isOpened is true', () => {
    component.instance().toggleOpened();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if toggleOpened sets isOpened correctly', () => {
    component.instance().toggleOpened();
    expect(component.instance().state.isOpened).toEqual(true);

    component.instance().toggleOpened();
    expect(component.instance().state.isOpened).toEqual(false);
  });

  it("should call toggleOpened after click on AddButton", () => {
    const spy = jest.spyOn(component.instance(), "toggleOpened");

    component.find('AddButton').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call toggleOpened after click on CancelButton", () => {
    const spy = jest.spyOn(component.instance(), "toggleOpened");

    component.setState({ isOpened: true });
    component.find('CancelButton').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  describe('add board form', () => {
    beforeEach(() => {
      component.setState({ isOpened: true });
    });

    it("should call toggleOpened if key code equal 27", () => {
      const spy = jest.spyOn(component.instance(), "toggleOpened");

      component.instance().handleKeyDown({ keyCode: 27 });

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call handleChange on boardTitle change with the correct params', () => {
      const spy = jest.spyOn(component.instance(), "handleChange");
      const mockEvent = {
        target: {
          name: "title",
          value: "Test title"
        }
      };

      component.find("Input").simulate("change", mockEvent);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith("boardTitle", "Test title");
    });

    it('should call handleChange on boardColor after click on color item', () => {
      const spy = jest.spyOn(component.instance(), "handleChange");

      component.find('ColorPicker').childAt(0).simulate('click');

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith("boardColor", "#2E7EAF");

      component.find('ColorPicker').childAt(1).simulate('click');

      expect(spy).toHaveBeenCalledTimes(2);
      expect(spy).toHaveBeenCalledWith("boardColor", "#00603d");
    });

    it('should call setState on boardTitle', () => {
      const expectedState = {
        boardTitle: "Test Title",
        boardColor: "",
        isOpened: true
      };

      component.instance().handleChange("boardTitle", "Test Title");

      expect(component.state()).toEqual(expectedState)
    });

    it('should call setState on boardColor', () => {
      const expectedState = {
        boardTitle: "",
        boardColor: "#D29034",
        isOpened: true
      };

      component.instance().handleChange("boardColor", "#D29034");

      expect(component.state()).toEqual(expectedState)
    });

    it('should render fontawesome icon if boardColor and color are the same', () => {
      component.setState({ boardTitle: 'Test Title', boardColor: '#D29034' });

      expect(toJson(component)).toMatchSnapshot();
    });

    it("should call handleSubmit with the correct params", () => {
      const spy = jest.spyOn(component.instance(), "handleSubmit");
      const mockEvent = {
        preventDefault: jest.fn()
      };

      component.setState({ boardTitle: 'Test Title', boardColor: '#D29034' });
      component.instance().handleSubmit(mockEvent);

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should return handleSubmit if title is empty", () => {
      const spy = jest.spyOn(component.instance(), "handleSubmit");
      const mockEvent = {
        preventDefault: jest.fn()
      };

      component.instance().handleSubmit(mockEvent);

      expect(spy).toReturn();
    });

    it("should dispatch an action after handleSubmit", () => {
      const expectedAction = {
        type: 'ADD_BOARD',
        payload: {
          boardTitle: 'Test Title',
          boardColor: '#D29034',
          boardId: expect.any(String),
        }
      };

      component.setState({ boardTitle: 'Test Title', boardColor: '#D29034' });
      component.find('SubmitButton').simulate('click');

      expect(mockAddBoard).toHaveBeenCalledTimes(1);
      expect(mockAddBoard).toHaveBeenCalledWith(expectedAction);
    });
  })
});
