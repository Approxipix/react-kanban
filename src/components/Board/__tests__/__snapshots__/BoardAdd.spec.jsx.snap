// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardAdd> add board form should render fontawesome icon if boardColor and color are the same 1`] = `
<ClickOutside
  toggleOpened={[Function]}
>
  <styled.div
    color="#D29034"
  >
    <form
      onSubmit={[Function]}
    >
      <Input
        autoFocus={true}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Add board title"
        spellCheck={false}
        type="text"
        value="Test Title"
      />
      <ColorPicker>
        <styled.div
          color="#2E7EAF"
          key="0"
          onClick={[Function]}
        />
        <styled.div
          color="#00603d"
          key="1"
          onClick={[Function]}
        />
        <styled.div
          color="#D29034"
          key="2"
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="check"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.div>
      </ColorPicker>
      <styled.div>
        <SubmitButton
          disabled={false}
          type="submit"
        >
          Create board
        </SubmitButton>
        <CancelButton
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="times"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </CancelButton>
      </styled.div>
    </form>
  </styled.div>
</ClickOutside>
`;

exports[`<BoardAdd> should render AddButton if isOpened is false 1`] = `
<AddButton
  onClick={[Function]}
>
  Create new board...
</AddButton>
`;

exports[`<BoardAdd> should render form if isOpened is true 1`] = `
<ClickOutside
  toggleOpened={[Function]}
>
  <styled.div
    color=""
  >
    <form
      onSubmit={[Function]}
    >
      <Input
        autoFocus={true}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Add board title"
        spellCheck={false}
        type="text"
        value=""
      />
      <ColorPicker>
        <styled.div
          color="#2E7EAF"
          key="0"
          onClick={[Function]}
        />
        <styled.div
          color="#00603d"
          key="1"
          onClick={[Function]}
        />
        <styled.div
          color="#D29034"
          key="2"
          onClick={[Function]}
        />
      </ColorPicker>
      <styled.div>
        <SubmitButton
          disabled={true}
          type="submit"
        >
          Create board
        </SubmitButton>
        <CancelButton
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="times"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </CancelButton>
      </styled.div>
    </form>
  </styled.div>
</ClickOutside>
`;
